ledger/mruaccts.go

```diff
@@ -0,0 +1,125 @@
+	pendingAccounts chan persistedAccountData
+	// log interface; used for logging the threshold event.
+	log logging.Logger
+	// pendingWritesWarnThreshold is the threshold beyond we would write a warning for exceeding the number of pendingAccounts entries
+	pendingWritesWarnThreshold int
+}
+
+// init initializes the mruAccounts for use.
+// thread locking semantics : write lock
+func (m *mruAccounts) init(log logging.Logger, pendingWrites int, pendingWritesWarnThreshold int) {
+	m.accountsList = list.New()
+	m.accounts = make(map[basics.Address]*list.Element)
+	m.pendingAccounts = make(chan persistedAccountData, pendingWrites)
+	m.log = log
+	m.pendingWritesWarnThreshold = pendingWritesWarnThreshold
+}
+
+// read the persistedAccountData object that the mruAccounts has for the given address.
+// thread locking semantics : read lock
+func (m *mruAccounts) read(addr basics.Address) (data persistedAccountData, has bool) {
+	if el := m.accounts[addr]; el != nil {
+		return el.Value.(persistedAccountData), true
+	}
+	return persistedAccountData{}, false
+}
+
+// flushPendingWrites flushes the pending writes to the main mruAccounts cache.
+// thread locking semantics : write lock
+func (m *mruAccounts) flushPendingWrites() {
+	pendingEntriesCount := len(m.pendingAccounts)
+	if pendingEntriesCount >= m.pendingWritesWarnThreshold {
+		m.log.Warnf("mruAccounts: number of entries in pendingAccounts(%d) exceed the warning threshold of %d", pendingEntriesCount, m.pendingWritesWarnThreshold)
+	}
+	for ; pendingEntriesCount > 0; pendingEntriesCount-- {
+		select {
+		case pendingAccountData := <-m.pendingAccounts:
+			m.write(pendingAccountData)
+		default:
+			return
+		}
+	}
+	return
+}
+
+// writePending write a single persistedAccountData entry to the pendingAccounts buffer.
+// the function doesn't block, and in case of a buffer overflow the entry would not be added.
+// thread locking semantics : no lock is required.
+func (m *mruAccounts) writePending(acct persistedAccountData) {
+	select {
+	case m.pendingAccounts <- acct:
+	default:
+	}
+}
+
+// write a single persistedAccountData to the mruAccounts cache.
+// when writing the entry, the round number would be used to determine if it's a newer
+// version of what's already on the cache or not. In all cases, the entry is going
+// to be promoted to the front of the list.
+// thread locking semantics : write lock
+func (m *mruAccounts) write(acctData persistedAccountData) {
+	if el := m.accounts[acctData.addr]; el != nil {
+		// already exists; is it a newer ?
+		existing := el.Value.(persistedAccountData)
+		if existing.before(&acctData) {
+			// we update with a newer version.
+			el.Value = acctData
+		}
+		m.accountsList.MoveToFront(el)
+	} else {
+		// new entry.
+		m.accounts[acctData.addr] = m.accountsList.PushFront(acctData)
+	}
+}
+
+// resize adjust the current size of the mruAccounts cache, by dropping the least
+// recenrly used entries.
```

Comment:
> dropping the least recenrly used

1. Typo in `recenrly`
2. The cache is MRU so why dropping LRU?

Reply:
1. fixed.
2. To maintain a MRU cache you drop the LRU ( i.e. all the ones that aren't recently used )

Reply:
I always thought MRU drops MRU.

Reply:
The way I understood it is that the LRU algorithm is used to determine the elements to drop from an MRU cache. In most cases, the MRU cache is not called the "MRU Cache" but just a "cache" in which case the removal logic is a "LRU cache cleanup". Given that I've explicitly called it a "MRU Cache", I don't think that the removal need to say "LRU".

Reply:
from some university [lecture](https://www.cs.cornell.edu/courses/cs4410/2015su/lectures/lec15-replacement.html):
>  Least recently used assumes that pages that were accessed recently are likely to be needed. Keep a timestamp of latest access, evict the page with the lowest timestamp. - Most recently used assumes that programs do not read the same addresses multiple times. For example, a media player will read a byte and then move on, never to read it again. As with LRU, keep a timestamp of latest access, but evict the page with the highest timestamp.

wikipedia [agrees](https://en.wikipedia.org/wiki/Cache_replacement_policies#Most_recently_used_(MRU))
> Most recently used (MRU)
Discards, in contrast to LRU, the most recently used items first. In findings presented at the 11th VLDB conference, Chou and DeWitt noted that "When a file is being repeatedly scanned in a [Looping Sequential] reference pattern, MRU is the best replacement algorithm."

In the implementation the position in the list is an analogue of access timestamp. Sounds like the cache needs to be renamed to LRU